tests/js/spec/stores/persistedStore.spec.jsx

```diff
@@ -0,0 +1,131 @@
+import {
+  PersistedStoreProvider,
+  usePersistedStoreCategory,
+} from 'sentry/stores/persistedStore';
+import {OrganizationContext} from 'sentry/views/organizationContext';
+
+describe('PersistedStore', function () {
+  it('provides, sets and deletes the persisted data category from client-state API', async function () {
+    const org = TestStubs.Organization();
+    MockApiClient.addMockResponse({
+      url: `/organizations/${org.slug}/client-state/`,
+      body: {
+        onboarding: {
+          test: 1,
+        },
+      },
+    });
+    OrganizationStore.onUpdate(org);
+    const wrapper = ({children}) => (
+      <PersistedStoreProvider>
+        <OrganizationContext.Provider value={org}>
+          {children}
+        </OrganizationContext.Provider>
+      </PersistedStoreProvider>
+    );
+
+    const {result, waitForNextUpdate} = reactHooks.renderHook(
+      () => usePersistedStoreCategory('onboarding'),
+      {wrapper}
+    );
+    await waitForNextUpdate();
+    const [state, setState] = result.current;
+
+    expect(state).toMatchObject({
+      test: 1,
+    });
+
+    // Set
+    const clientStateUpdate = MockApiClient.addMockResponse({
+      url: `/organizations/${org.slug}/client-state/onboarding/`,
+      method: 'PUT',
+    });
+
+    act(() => {
+      setState({test: 2});
+    });
+
+    const [state2, setState2] = result.current;
+    expect(state2).toMatchObject({
+      test: 2,
+    });
+    expect(clientStateUpdate).toHaveBeenCalledWith(
+      `/organizations/${org.slug}/client-state/onboarding/`,
+      expect.objectContaining({
+        method: 'PUT',
+        data: {
+          test: 2,
+        },
+      })
+    );
+
+    // Delete
+    const clientStateDelete = MockApiClient.addMockResponse({
+      url: `/organizations/${org.slug}/client-state/onboarding/`,
+      method: 'DELETE',
+    });
+    act(() => {
+      setState2(null);
+    });
+    const [state3, _] = result.current;
+    expect(state3).toBe(null);
+    expect(clientStateDelete).toHaveBeenCalledWith(
+      `/organizations/${org.slug}/client-state/onboarding/`,
+      expect.objectContaining({
+        method: 'DELETE',
+      })
+    );
+    MockApiClient.clearMockResponses();
+  });
+  it('behaves nicely when state not loaded', async function () {
+    const org = TestStubs.Organization();
+    MockApiClient.addMockResponse({
+      url: `/organizations/${org.slug}/client-state/`,
+      status: 500,
+    });
+    OrganizationStore.onUpdate(org);
+    const wrapper = ({children}) => (
+      <PersistedStoreProvider>
+        <OrganizationContext.Provider value={org}>
+          {children}
+        </OrganizationContext.Provider>
+      </PersistedStoreProvider>
+    );
+
+    const {result, waitForNextUpdate} = reactHooks.renderHook(
+      () => usePersistedStoreCategory('onboarding'),
+      {wrapper}
+    );
+    await waitForNextUpdate();
+    const [state, setState] = result.current;
+    expect(state).toBe(null);
```

Comment:
If you want to avoid things like await waitForNextUpdate, there is a nicer way to test this that doesnt rely on manually awaiting timers and is more bulletproof.

```suggestion
    const TestComponent = () => {
       const [state] = usePersistedStoreCategory('onboarding')
       return <Fragment>{state.someThingYouCanAssertOn}</Fragment>
    }
    
    render(<TestComponent/>), {wrapper})
    expect(await screen.findByText(/someThingYouCanAssertOne/)).toBeInTheDocument())
```
I would not test that initialState is null, that sounds like an implementation detail tbh

Reply:
imo this is fine, I did this because `renderHook` is designed for this and using `render` here seems more like a hack to me. `renderHook` is also recommended here: https://react-hooks-testing-library.com/usage/advanced-hooks

Reply:
Yeah, the only problem with waitForNextUpdate is that if you change something internally, this may fail because it'll always wait just 1 tick whereas findBy will wait for the thing to show up no matter how long it takes.